backend/src/main/java/gov/cdc/usds/simplereport/config/DataLoaders.java

```diff
@@ -0,0 +1,64 @@
+package gov.cdc.usds.simplereport.config;
+
+import static java.util.concurrent.CompletableFuture.supplyAsync;
+
+import gov.cdc.usds.simplereport.db.model.PatientPreferences;
+import gov.cdc.usds.simplereport.db.model.TestEvent;
+import gov.cdc.usds.simplereport.db.repository.PatientPreferencesRepository;
+import gov.cdc.usds.simplereport.db.repository.TestEventRepository;
+import java.util.Map;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import org.dataloader.DataLoader;
+import org.dataloader.DataLoaderRegistry;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DataLoaders {
+  private final PatientPreferencesRepository patientPreferencesRepository;
+  private final TestEventRepository testEventRepository;
+  public static final String PATIENT_LAST_TEST = "patientLastTestLoader";
+  public static final String PATIENT_PREFERENCES = "patientPreferencesLoader";
+
+  public DataLoaders(
+      PatientPreferencesRepository patientPreferencesRepository,
+      TestEventRepository testEventRepository) {
+    this.patientPreferencesRepository = patientPreferencesRepository;
+    this.testEventRepository = testEventRepository;
+  }
+
+  private DataLoader<UUID, PatientPreferences> createPatientPreferencesLoader() {
+    return new DataLoader<>(
+        patientIds ->
+            supplyAsync(
+                () -> {
+                  Map<UUID, PatientPreferences> found =
+                      patientPreferencesRepository.findAllByPersonInternalIdIn(patientIds).stream()
+                          .collect(Collectors.toMap(PatientPreferences::getInternalId, s -> s));
+                  return patientIds.stream()
+                      .map(p -> found.getOrDefault(p, PatientPreferences.DEFAULT))
+                      .collect(Collectors.toList());
+                }));
+  }
+
+  private DataLoader<UUID, TestEvent> createTestEventDataLoader() {
+    return new DataLoader<>(
+        patientIds ->
+            supplyAsync(
+                () -> {
+                  Map<UUID, TestEvent> found =
+                      testEventRepository.findLastTestsByPatient(patientIds).stream()
+                          .collect(Collectors.toMap(TestEvent::getPatientInternalID, s -> s));
+                  return patientIds.stream()
+                      .map(te -> found.getOrDefault(te, null))
+                      .collect(Collectors.toList());
+                }));
+  }
+
+  public DataLoaderRegistry buildRegistry() {
+    DataLoaderRegistry dataLoaderRegistry = new DataLoaderRegistry();
+    dataLoaderRegistry.register(PATIENT_PREFERENCES, createPatientPreferencesLoader());
+    dataLoaderRegistry.register(PATIENT_LAST_TEST, createTestEventDataLoader());
```

Comment:
Could we be using Spring DI to make this pool of loaders easier to expand in the future? I.e., if each loader were a component or bean, the gathered set could be injected into a `DataLoaderRegistry` component.

Reply:
Yeah, that's basically what I was thinking. Not sure if Spring has an equivalent to Dagger's `@IntoMap`, but I would imagine this looking something like:

```
    @Bean
    @IntoMap
    Map.Entry<String, DataLoader> createPatientPreferencesLoader() { ... }
    
    @Bean
    @IntoMap
    Map.Entry<String, DataLoader> createTestEventDataLoader() { ... }
    
    @Bean
    DataLoaderRegistry createDataLoaderRegistry(Map<String, DataLoader> loaders) { ... }
```

Reply:
It looks like we could do it like so...

```
@Autowired Map<String, DataLoader> dataLoaders;
```

And separate the individual loaders into `@Bean`s that `extend DataLoader`. Source: https://stackoverflow.com/a/32740130

I'll try it and report back

Reply:
Okay so I realize that this is doubly approved, but the changes that I made (which resolve this concern) might be terrible in their own way? Would love another quick glance if it's not too much trouble 😬 @jeskew-gov @benwarfield-usds

Reply:
**Update:** Jonny addressed the one contract-breaking concern that these changes introduced, and I've reworked it to resolve them. Feeling really confident about this work now, so I'm going to let it sit for a minute but will plan to merge it tomorrow if there's no further discussion.